File: tests/samples/Perceptron/Neuron.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define Dataset 1 // 0 - And | 1 - Or
#define LearningRate 0.1f

#define max(a, b) ((a) > (b) ? (a) : (b))

int Epoch = 1;

float Error;
float ErrorOccurred;

typedef struct {
    float *Weights;
    
    int InputAmount;
    float Bias;
    float Output;
} Neuron;

float Step(float Value) {
    return Value >= 0 ? 1.0f : 0.0f;
}

Neuron *CreateNeuron(int InputAmount) {
    Neuron *_Neuron = malloc(sizeof(Neuron));

    _Neuron -> Weights = malloc(InputAmount * sizeof(float));
    _Neuron -> InputAmount = InputAmount;
    _Neuron -> Bias = 0;
    
    for (int i = 0; i < InputAmount; i++) {
        _Neuron -> Weights[i] = 0;
    }

    return _Neuron;
}

float ForwardPass(Neuron *_Neuron, float *Inputs, float ExpectedOutput) {
    float Sum = 0.0f;

    for (int i = 0; i < _Neuron -> InputAmount; i++) {
        Sum += _Neuron -> Weights[i] * Inputs[i];
    }

    Sum += _Neuron -> Bias;
    _Neuron -> Output = Step(Sum);

    Error = ExpectedOutput - _Neuron -> Output;
    if (Error != 0)
        ErrorOccurred = 1;

    printf("Expected output: %.1f\n", ExpectedOutput);
    printf("Error: %.2f\n", Error);

    for (int i = 0; i < 2; i++) {
        _Neuron -> Weights[i] += LearningRate * Error * Inputs[i];
    }

    _Neuron -> Bias += LearningRate * Error;

    printf("Weighted sum (z): %.2f\n", Sum);

    return _Neuron -> Output;
}

int main() {
    float Inputs[4][2] = {
        {0, 0}, {0, 1}, {1, 0}, {1, 1}
    };

    float ExpectedOutputs[2][4] = {
        {0, 0, 0, 1}, {0, 1, 1, 1}
    };

    Neuron *_Neuron = CreateNeuron(2);

    int Activations = 0;

    printf("====================\n\n");

    do {
        ErrorOccurred = 0;
        Activations = 0;

        printf("Epoch: %i\n\n", Epoch);

        for (int i = 0; i <  4; i++) {
            float Output = ForwardPass(_Neuron, Inputs[i], ExpectedOutputs[Dataset][i]);
            if (_Neuron -> Output == 1.0f)
                Activations++;

            printf("Inputs: [%.0f, %.0f]\n", Inputs[i][0], Inputs[i][1]);
            printf("Bias: %.2f\n", _Neuron -> Bias);
            printf("Output: %.2f\n\n", Output);
        }

        printf("====================\n\n");

        Epoch++;
    } while (ErrorOccurred == 1);

    printf("Total Epoch's: %i\n", Epoch - 1);
    printf("Final Guess: %s\n", Activations == 1 ? "And" : "Or");

    free(_Neuron -> Weights);
    free(_Neuron);
}

File: tests/samples/Perceptron/Neuron.lua
Dataset = 1 -- 0 : And | 1 : Or
LearningRate = 0.1

Epoch = 1
Activations = 0

Error = 0
ErrorOccurred = false

function Max(a, b)
    if a > b then
        return a
    else
        return b
    end
end

function Step(Value)
    if Value >= 0 then
        return 1.0
    else
        return 0.0
    end
end

function CreateNeuron(InputAmount)
    local Neuron = {}

    Neuron.Weights = {}

    for i = 1, InputAmount do
        Neuron.Weights[i] = 0.0
    end

    Neuron.Bias = 0.0
    Neuron.Output = 0.0

    return Neuron
end

function ForwardPass(Neuron, Inputs, ExpectedOutput)
    local Sum = 0.0

    for i = 1, #Inputs do
        Sum = Sum + Inputs[i] * Neuron.Weights[i]
    end

    Sum = Sum + Neuron.Bias
    Neuron.Output = Step(Sum)

    Error = ExpectedOutput - Neuron.Output
    if Error ~= 0 then
        ErrorOccurred = true
    end

    print("Expected: " .. ExpectedOutput)
    print("Error: " .. Error)

    for i = 1, #Inputs do
        Neuron.Weights[i] = Neuron.Weights[i] + LearningRate * Error * Inputs[i]
    end

    Neuron.Bias = Neuron.Bias + LearningRate * Error

    print("Weighted sum (z): " .. Sum)

    return Neuron.Output
end

local Inputs = {
    {0.0, 0.0},
    {0.0, 1.0},
    {1.0, 0.0},
    {1.0, 1.0}
}

local ExpectedOutputs = {
    {0.0, 0.0, 0.0, 1.0},
    {0.0, 1.0, 1.0, 1.0}
}

local Neuron = CreateNeuron(2)

print("====================\n")

while true do
    print("Epoch: " .. Epoch)

    ErrorOccurred = false

    for i = 1, #Inputs do
        print("")

        local Output = ForwardPass(Neuron, Inputs[i], ExpectedOutputs[Dataset + 1][i])

        print("Input: [" .. Inputs[i][1] .. ", " .. Inputs[i][2] .. "]")
        print("Bias: " .. Neuron.Bias)
        print("Output: " .. Output)
    end

    print("\n====================\n")

    if not ErrorOccurred then
        break
    end
    
    Epoch = Epoch + 1
end

print("Total Epochs: " .. Epoch)

if Activations == 1 then
    print("Final Guess: AND")
else
    print("Final Guess: OR")
end
